File: javascript/tests/pinning.test.js
/**
 * Tests for TOFU key pinning.
 */

import { test, describe } from 'node:test';
import assert from 'node:assert';
import { KeyPinStore, PinningResult, checkPinning } from '../src/pinning.js';

function makeTestJwk(kid, x) {
    return {
        kid,
        kty: 'EC',
        crv: 'P-256',
        x,
        y: 'test-y',
        use: 'sig',
    };
}

describe('KeyPinStore', () => {
    test('first use pins key', () => {
        const store = new KeyPinStore();
        const jwk = makeTestJwk('key-1', 'x-value-1');
        const result = store.checkAndPin('example.com', jwk);
        assert.strictEqual(result, PinningResult.FIRST_USE);

        const result2 = store.checkAndPin('example.com', jwk);
        assert.strictEqual(result2, PinningResult.MATCHED);
    });

    test('key change detected', () => {
        const store = new KeyPinStore();
        const jwk1 = makeTestJwk('key-1', 'x-value-1');
        store.checkAndPin('example.com', jwk1);

        const jwk2 = makeTestJwk('key-2', 'x-value-2');
        const result = store.checkAndPin('example.com', jwk2);
        assert.strictEqual(result, PinningResult.CHANGED);
    });

    test('addKey allows rotation', () => {
        const store = new KeyPinStore();
        const jwk1 = makeTestJwk('key-1', 'x-value-1');
        store.checkAndPin('example.com', jwk1);

        const jwk2 = makeTestJwk('key-2', 'x-value-2');
        store.addKey('example.com', jwk2);

        assert.strictEqual(store.checkAndPin('example.com', jwk1), PinningResult.MATCHED);
        assert.strictEqual(store.checkAndPin('example.com', jwk2), PinningResult.MATCHED);
    });

    test('JSON roundtrip', () => {
        const store = new KeyPinStore();
        const jwk = makeTestJwk('key-1', 'x-value-1');
        store.checkAndPin('example.com', jwk);

        const json = store.toJson();
        const store2 = KeyPinStore.fromJson(json);
        assert.ok(store2.getDomain('example.com'));
        assert.strictEqual(store2.getDomain('example.com').pinned_keys.length, 1);
    });

    test('different domains are independent', () => {
        const store = new KeyPinStore();
        const jwk1 = makeTestJwk('key-1', 'x-value-1');
        const jwk2 = makeTestJwk('key-2', 'x-value-2');

        store.checkAndPin('a.com', jwk1);
        store.checkAndPin('b.com', jwk2);

        assert.strictEqual(store.checkAndPin('a.com', jwk1), PinningResult.MATCHED);
        assert.strictEqual(store.checkAndPin('b.com', jwk2), PinningResult.MATCHED);
        assert.strictEqual(store.checkAndPin('a.com', jwk2), PinningResult.CHANGED);
    });
});

describe('checkPinning', () => {
    test('throws on key change', () => {
        const store = new KeyPinStore();
        const jwk1 = makeTestJwk('key-1', 'x-value-1');
        store.checkAndPin('example.com', jwk1);

        const jwk2 = makeTestJwk('key-2', 'x-value-2');
        assert.throws(() => checkPinning(store, 'example.com', jwk2), /changed/i);
    });
});

File: go/pkg/pinning/pinning_test.go
package pinning

import (
	"testing"

	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

func makeJWK(kid, x string) *types.JWK {
	return &types.JWK{
		Kid: kid, Kty: "EC", Crv: "P-256",
		X: x, Y: "test-y", Use: "sig",
	}
}

func TestFirstUsePinsKey(t *testing.T) {
	store := NewKeyPinStore()
	j := makeJWK("k", "x1")
	if r := store.CheckAndPin("example.com", j); r != ResultFirstUse {
		t.Fatalf("first call should be FirstUse, got %v", r)
	}
	if r := store.CheckAndPin("example.com", j); r != ResultMatched {
		t.Fatalf("second call should be Matched, got %v", r)
	}
}

func TestKeyChangeDetected(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("example.com", makeJWK("k", "x1"))
	if r := store.CheckAndPin("example.com", makeJWK("k", "x2")); r != ResultChanged {
		t.Fatalf("changed key should be Changed, got %v", r)
	}
}

func TestAddKeyAllowsRotation(t *testing.T) {
	store := NewKeyPinStore()
	j1 := makeJWK("k1", "x1")
	j2 := makeJWK("k2", "x2")
	store.CheckAndPin("example.com", j1)
	store.AddKey("example.com", j2)
	if r := store.CheckAndPin("example.com", j2); r != ResultMatched {
		t.Fatalf("rotated key should match, got %v", r)
	}
}

func TestPinJSONRoundTrip(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("example.com", makeJWK("k", "x"))
	data, err := store.MarshalJSON()
	if err != nil {
		t.Fatal(err)
	}
	store2 := NewKeyPinStore()
	if err := store2.LoadFromJSON(data); err != nil {
		t.Fatal(err)
	}
	if pd := store2.GetDomain("example.com"); pd == nil || len(pd.PinnedKeys) != 1 {
		t.Fatalf("after roundtrip: %+v", pd)
	}
}

func TestCheckPinningError(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("example.com", makeJWK("k", "x1"))
	if _, err := CheckPinning(store, "example.com", makeJWK("k", "x2")); err == nil {
		t.Fatal("expected pin error")
	}
}

func TestDifferentDomainsIndependent(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("a.com", makeJWK("k", "xa"))
	store.CheckAndPin("b.com", makeJWK("k", "xb"))
	if r := store.CheckAndPin("a.com", makeJWK("k", "xb")); r != ResultChanged {
		t.Fatalf("cross-domain: got %v", r)
	}
}
